File: frontend/src/services/userService.ts
import axios from 'axios';
import type { User } from '../types/User';

const BASE_URL = import.meta.env.VITE_API_BASE_URL;

if (!BASE_URL) {
    throw new Error('VITE_API_BASE_URL environment variable is not set');
}

const API_URL = `${BASE_URL}/api/users`;

export const userService = {
    createUser: async (user: User) => {
        const response = await axios.post(API_URL, user);
        return response.data;
    },

    getAllUsers: async () => {
        const response = await axios.get(API_URL);
        return response.data;
    },
    
    getAllActiveUsers: async () => {
        const response = await axios.get(`${API_URL}/active`);
        return response.data;
    },
    
    getUserById: async (id: string) => {
        const response = await axios.get(`${API_URL}/${id}`);
        return response.data;
    },

    updateUser: async (id: string, user: User) => {
        const response = await axios.put(`${API_URL}/${id}`, user);
        return response.data;
    },

    softDeleteUser: async (id: string) => {
        const response = await axios.put(`${API_URL}/${id}/soft`);
        return response.data;
    },

    hardDeleteUser: async (id: string) => {
        const response = await axios.delete(`${API_URL}/${id}/hard`);
        return response.data;
    }
};

File: backend/src/main/java/com/example/todo/functions/userCrud/service/UserService.java
package com.example.todo.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.todo.entity.User;
import com.example.todo.repository.UserRepository;

@Service
@Transactional
public class UserService {  

    @Autowired
    private UserRepository userRepository;
    
    //retrieve all users (including soft-deleted)
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    //retrieve all active users (not soft-deleted)
    public List<User> getAllActiveUsers() {
        return userRepository.findAllActive();
    }
    
    //retrieve a user by id
    public User getUserById(String id) {
        return userRepository.findById(id).orElse(null);
    }

    //create a new user
    public User createUser(User user) {
        if (userRepository.existsByEmail(user.getEmail())) {
            return null;
        }

        user.setId(null); 
        return userRepository.save(user);
    }

    //update an existing user
    public User updateUser(String id, User user) {
        if (userRepository.existsById(id)) {
            user.setId(id);
            return userRepository.save(user);
        }
        return null;
    }

    //soft delete a user by id (existing user)
    public void softDeleteUser(String id) {
        if (userRepository.existsById(id)) {
            userRepository.softDeleteUserById(id);
        }
    }

    //hard delete a user by id (existing user)
    public void hardDeleteUser(String id) {
        if (userRepository.existsById(id)) {
            userRepository.deleteById(id);
        }
    }

}
